### PR TITLE
PostgreSQL & ShardLayout

### PR DESCRIPTION
**Pull request by kobayurii**
As our data size continues to grow, managing the load has become increasingly challenging, leading to several issues:

- Slowing down of contracts due to state provision in contract runtime [#150](https://github.com/near/read-rpc/pull/link)
- View calls are either too slow or return HTTP/502 (Bad Gateway) errors [#193](https://github.com/near/read-rpc/pull/link)
- Performance problems featuring near-vm [#236](https://github.com/near/read-rpc/pull/link)

After extensive discussions and brainstorming, we believe these issues can be mitigated by transitioning to a more traditional relational database management system, specifically PostgreSQL.

**In This PR**

1. Improved or refactored the database crate to align with the shard layout from nearcore [#289](https://github.com/near/read-rpc/pull/link)
2. Redesigned and designed the missing parts of the PostgreSQL schemas, respecting the shard layout [#290](https://github.com/near/read-rpc/pull/link)

**By implemented these changes, we anticipate the following benefits:**

1. Increased request handling capacity due to the relational DB's ability to fetch data in batches effectively.
2. The ability to implement the long-standing query.access_key_list method (reference: Redesign data collection and storage for query.view_access_key_list RPC method [#187](https://github.com/near/read-rpc/pull/link)).

### CODE REVIEW

**Comment by kobayurii** [comment]
@race-of-sloths please, include my PR in the Race

**Comment by race-of-sloths** [comment]
@kobayurii Thank you for your contribution! Your pull request is now a part of the [Race of Sloths](https://race-of-sloths.com)!
New Sloth joined the Race! Welcome!

[<picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://badge.race-of-sloths.com/kobayurii?type=bot&theme=dark">
    <source media="(prefers-color-scheme: light)" srcset="https://badge.race-of-sloths.com/kobayurii?type=bot&theme=light">
    Shows profile picture for the author of the PR
</picture>](https://race-of-sloths.com/profile/kobayurii)


<details><summary>Current status: <i>waiting for scoring</i></summary>

We're waiting for maintainer to score this pull request with `@race-of-sloths score [0,1,2,3,5,8,13]` command
</details><details><summary>What is the Race of Sloths</summary>
<p>
Race of Sloths is a friendly competition where you can participate in challenges and compete with other open-source contributors within your normal workflow

For contributors:

- Tag @race-of-sloths inside your pull requests
- Wait for the maintainer to review and score your pull request
- Check out your position in the [Leaderboard](https://race-of-sloths.com/leaderboard)
- Keep weekly and monthly streaks to reach higher positions
- Boast your contributions with a dynamic picture of your [Profile](https://race-of-sloths.com/profile/kobayurii)

For maintainers:
- Score pull requests that participate in the Race of Sloths
- Engage contributors with fair scoring and fast responses so they keep their streaks
- Promote the Race to the point where the Race starts promoting you
- Grow the community of your contributors

Feel free to check [our website](https://race-of-sloths.com) for additional details!
</p>
</details> 

<details>
<summary>Bot commands</summary>

- For contributors
  - **Include a PR:** `@race-of-sloths include` to enter the Race with your PR
- For maintainers:
  - **Assign points:** `@race-of-sloths score [1/2/3/5/8/13]` to award points based on your assessment.
  - **Reject this PR:** `@race-of-sloths exclude` to send this PR back to the drawing board.
  - **Exclude repo:** `@race-of-sloths pause` to stop bot activity in this repo until `@race-of-sloths unpause` command is called

</details>

**Review by frolvanya — APPROVED** [review]

File: database/src/postgres/rpc_server.rs
Comment (khorolets) [inline database/src/postgres/rpc_server.rs:459]: Please comment here about our chosen approach: send a query to all shards, expecting only one to result in data.

File: database/src/postgres/state_indexer.rs
Comment (khorolets) [inline database/src/postgres/state_indexer.rs:259]: In such error messages would be great to include the shard id we tried to simplify the debugging

File: rpc-server/src/modules/queries/methods.rs
Comment (khorolets) [inline rpc-server/src/modules/queries/methods.rs:647]: It is time to have a look at this TODO I guess :)

**Review by khorolets — CHANGES_REQUESTED** [review]
That's a lot of work! Awesome work! I left comments at all the places that caught my eye. Please, point me to any specific parts you would like my extra attention.

**Comment by khorolets** [comment]
@race-of-sloths score 13

**Comment by race-of-sloths** [comment]
### 🌟 Score recorded!

@khorolets, thank you for scoring this pull request in the Race of Sloths!

File: database/src/postgres/mod.rs
Comment (kobayurii) [reply database/src/postgres/mod.rs:10]: we serialize this struct into `bytes` and do `hex:encode` to get `next_page token` to fetch state by pages.

**Comment by frolvanya** [comment]
@kobayurii Did the the latest commit fix an issue with contract calls? Now everything is ready for the load test?

**Comment by kobayurii** [comment]
> @kobayurii Did the the latest commit fix an issue with contract calls? Now everything is ready for the load test?

Yes, the latest changes have fixed the problem. We are now ready for the test.

**Comment by race-of-sloths** [comment]
### ❌ PR inactive
This PR has been inactive for two weeks. It's now marked as stale. 
>[!TIP]
If you think it's a mistake, you can restart the bot with `include` command to proceed 🔄

**Comment by gilxgil** [comment]
hi @kobayurii we're interested in running this for RPC Nodes, is this planned for main?

**Comment by kobayurii** [comment]
hi @gilxgil, yes, this will be merged into maine soon, we are doing final testing and debugging. You can try it out here https://v2.rpc.mainnet.near.org

**Comment by kobayurii** [comment]
@race-of-sloths please, include my PR in the Race

**Pull request merged**
